### PR TITLE
Generalize factory patterns for use outside of NSValue

### PR DESCRIPTION
**Pull request by phausler**

### CODE REVIEW

**Comment by phausler** [comment]
cc @norio-nomura, after looking at this a bit further I think we can do some really neat things. Other than the bare init case did you find any other caveats worth mentioning?

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

File: Foundation/NSNumberBridge.swift [hunk 1]
```diff
@@ -0,0 +1,698 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+
+#if DEPLOYMENT_RUNTIME_SWIFT
```
Comment (phausler) [inline Foundation/NSNumberBridge.swift]: This is now shareable to the Darwin overlay

**Comment by norio-nomura** [comment]
Wow! ❤️

**Comment by phausler** [comment]
oh I have more incoming. Prior to your approach we had two major differentials between the general behavior of the objc version of Foundation and swift-corelibs-foundation; factory patterns and bridging. This method effectively catches 90% of the factory pattern side of things. I have a version for NSString and NSMutableString an I am currently working on getting a few more types.

Overall this approach will vastly improve performance for swift-corelibs-foundation and dramatically get closer to the expected behavior for abstract base classes.

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

**Comment by alblue** [comment]
@phausler can you rebase this onto the latest? In particular, the project.pbxproj has recently had conflicts (which @ianpartridge resolved elsewhere) so would be good to make sure this doesn't conflict.

File: CoreFoundation/Base.subproj/CFInternal.h [hunk 2]
```diff
@@ -187,7 +187,7 @@ CF_PRIVATE CFIndex __CFActiveProcessorCount();
 #if defined(DEBUG)
     #define CFAssert(cond, prio, desc) do { if (!(cond)) { CFLog(prio, CFSTR(desc)); /* HALT; */ } } while (0)
     #define CFAssert1(cond, prio, desc, a1) do { if (!(cond)) { CFLog(prio, CFSTR(desc), a1); /* HALT; */ } } while (0)
-    #define CFAssert2(cond, prio, desc, a1, a2) do { if (!(cond)) { CFLog(prio, CFSTR(desc), a1, a2); /* HALT; */ } } while (0)
+    #define CFAssert2(cond, prio, desc, a1, a2) do { if (!(cond)) { CFLog(prio, CFSTR(desc), a1, a2);  HALT; } } while (0)
```
Comment (alblue) [inline CoreFoundation/Base.subproj/CFInternal.h]: Why is the HALT uncommented here?

File: Foundation/NSCFStream.swift [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+internal final class _NSCFInputStream : InputStream {
+    deinit {
+        if let obj = _CFReadStreamGetClient(_unsafeReferenceCast(self, to: CFReadStream.self)) {
+            _streamDelegates[obj] = nil
+        }
+    }
+    
+
```
Comment (alblue) [inline Foundation/NSCFStream.swift]: Minor nit, but there's two empty lines here where one would do.

File: Foundation/NSDictionary.swift [hunk 4]
```diff
@@ -260,25 +260,25 @@ open class NSDictionary : NSObject, NSCopying, NSMutableCopying, NSSecureCoding,
...
+                line += descriptionByType
             } else {
-                line.append("\(key)")
+                line += "\(key)"
             }
-            line.append("\"")
-            line.append(" = ")
-            line.append("\"")
+            line += "\""
+            line += " = "
```
Comment (alblue) [inline Foundation/NSDictionary.swift]: Given that we're appending these three constants together, is there a reason we don't just have them all in one constant `\" = \"` instead?

File: Foundation/NSObjCRuntime.swift [hunk 5]
```diff
@@ -206,6 +206,10 @@ internal func NSInvalidArgument(_ message: String, method: String = #function, f
     fatalError("\(method): \(message)", file: file, line: line)
 }
 
+internal func _NSMethodExceptionProem(_ obj: NSObject, _ cmd: StaticString = #function) -> String {
```
Comment (alblue) [inline Foundation/NSObjCRuntime.swift]: What does the `Proem` mean here?

File: Foundation/NSString.swift [hunk 6]
```diff
@@ -1,714 +1,948 @@
...
+  
         public static let caseInsensitive = CompareOptions(rawValue: 1)
         public static let literal = CompareOptions(rawValue: 2)
         public static let backwards = CompareOptions(rawValue: 4)
         public static let anchored = CompareOptions(rawValue: 8)
         public static let numeric = CompareOptions(rawValue: 64)
         public static let diacriticInsensitive = CompareOptions(rawValue: 128)
         public static let widthInsensitive = CompareOptions(rawValue: 256)
-        public static let forcedOrdering = CompareOptions(rawValue: 512)
+        public static var forcedOrdering = CompareOptions(rawValue: 512)
```
Comment (alblue) [inline Foundation/NSString.swift]: Can this really change? Why is it a global `var`?

File: Foundation/NSString.swift [hunk 7]
```diff
@@ -1149,7 +1455,7 @@ extension NSString {
         if !getBytes(nil, maxLength: Int.max - 1, usedLength: &numBytes, encoding: enc, options: [], range: theRange, remaining: nil) {
             throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.fileWriteInapplicableStringEncoding.rawValue, userInfo: [
                 NSURLErrorKey: dest,
-            ])
+                ])
```
Comment (alblue) [inline Foundation/NSString.swift]: Is this indentation expected? I thought it looked better before, where it was aligned with the 'throw'

File: Foundation/NSString.swift [hunk 8]
```diff
@@ -1159,7 +1465,7 @@ extension NSString {
             if !getBytes(mutableBytes, maxLength: numBytes, usedLength: &used, encoding: enc, options: [], range: theRange, remaining: nil) {
                 throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.fileWriteUnknown.rawValue, userInfo: [
                     NSURLErrorKey: dest,
-                ])
+                    ])
```
Comment (alblue) [inline Foundation/NSString.swift]: I would have thought the alignment would be better if aligned with the `throw`

File: TestFoundation/TestNSString.swift [hunk 9]
```diff
@@ -1103,37 +1103,37 @@ class TestNSString : XCTestCase {
...
 
-        let surrogatePairSuffix = NSString(string: "Hurray🎉")
-        XCTAssertEqual(surrogatePairSuffix.substring(with: NSMakeRange(0, 7)), "Hurray�")
-
-        let surrogatePairPrefix = NSString(string: "🐱Cat")
-        XCTAssertEqual(surrogatePairPrefix.substring(with: NSMakeRange(1, 4)), "�Cat")
-
-        let singleChar = NSString(string: "😹")
-        XCTAssertEqual(singleChar.substring(with: NSMakeRange(0,1)), "�")
+//        let surrogatePairSuffix = NSString(string: "Hurray🎉")
```
Comment (alblue) [inline TestFoundation/TestNSString.swift]: Were these test failures failing, or no longer relevant? If they're broken, do we have a bug to fix them again?

**Comment by phausler** [comment]
There are a number of other changes I have pending for this - namely it does not yet work fully on Linux, once I get back to the office next week I will revise those changes onto master and attempt to get this landed

File: CoreFoundation/Base.subproj/CFInternal.h (hunk 2 above)
Comment (phausler) [inline CoreFoundation/Base.subproj/CFInternal.h]: I was using it to catch some failures in bridge casts

File: Foundation/NSObjCRuntime.swift (hunk 5 above)
Comment (phausler) [inline Foundation/NSObjCRuntime.swift]: this is taken from the actual Foundation; https://www.merriam-webster.com/dictionary/proem it is a preliminary comment

File: Foundation/NSString.swift (hunk 6 above)
Comment (phausler) [inline Foundation/NSString.swift]: hmm probably shouldn't be

File: TestFoundation/TestNSString.swift (hunk 9 above)
Comment (phausler) [inline TestFoundation/TestNSString.swift]: these tests iirc don't actually work on Darwin, will have to double check

**Comment by phausler** [comment]
@swift-ci please test

**Comment by alblue** [comment]
@phausler there was a lot of work done on this branch, but it's got out of step. Can you rebase and re-run the tests please?

**Comment by alblue** [comment]
@phausler This pull request has gone stale, and there are a number of conflicts which prevent it moving forward at this stage. I'm going to tentatively close it, and we can re-open in the future once the conflicts have been fixed.

**Pull request closed without merging**